### ISSUE TITLE
Non-MARC import needs to handle subtitles correctly

### ISSUE THREAD

**Issue by hornc** [internetarchive/openlibrary#2311]
68% of seen BWB published in 2019 have subtitles denoted by colons in the title, currently subtitles are one of the areas not handled very well when it comes to identifying existing items.

MARC import has the facility to split out Title into title + subtitle.

Non-MARC import does not appear to do this. Recent example: https://openlibrary.org/books/OL27268600M/Whisper_Network_A_Novel  A Novel looks like it should have been made a subtitle if it had been imported from MARC.

The task is to consolidate Non-MARC import process subtitle handling so at least our data is in consistent form. Consistent results on each import is the first step.

Additional tasks will be to review the title matching algorithm to see whether it, or our data model, can be improved.

relates to #2220

**Comment by hornc** [comment]
complex parens not handled 100%: https://dev.openlibrary.org/books/OL27322545M/Last_Days_at_Hot_Slit

**Comment by cdrini** [comment]
Hmmm, I'm not sure we should be stripping out the parens. It might contain useful information. I think I'd rather editors decide what to do with it.

Otherwise, this regex should do the trick: `replace('\(.*\)', '')` https://regexr.com/4lqvm

**Comment by hornc** [comment]
@cdrini if you look at the notes field of that example you'll see I am adding the full original title in the edition notes field for that purpose.

From these sources, the parens are generally not part of the title, so keeping them there breaks any title matching. I'm trying to clean titles, but retain 'other' info in notes.

From what I've seen the parens text is usually a series name, and / or volume info, but it is often other info such as awards.

Once enough of these are processed I was imagining an automated pass to check these notes fields and update series etc using what is available.

**Comment by hornc** [comment]
Checking your regex, I was deliberatly trying to make sure I removed every group in parens, but didn't strip text between. I have seen titles with sequential parens, but I don't recall one that did have title text between them. I'll see if I can find examples.

I think I'd prefer a regex that can cleanly remove nested parens, but that may be overkill.

**Comment by tfmorris** [comment]
The Amazon [page](https://www.amazon.com/gp/product/1635900808) explicitly lists the series for the first book as the string in the (outer) parens, so we definitely want to strip that (and most/all other parenthetical expressions in a title).

**Comment by seabelis** [comment]
I'm not sure about bwb, but there's a lot of title stuffing on amazon. Would prefer to strip it out, at least from their data, if possible. Proper subtitles are usually included after a colon.

**Comment by cdrini** [comment]
> I think I'd prefer a regex that can cleanly remove nested parens, but that may be overkill.

That's actually not possible with most implementations of regex. (See http://rachbelaid.com/recursive-regular-experession/ ). I'd be content with removing parenthetical information _if it's at the end of the title (or subtitle)_.

**Comment by tfmorris** [comment]
I don't think we need any fancy recursion - just removing everything between the first left paren and the last right paren should be fine.

**Comment by hornc** [comment]
I've looked for some pathological examples and come up with these:

* Secrets of Adobe(r) Acrobat(r) 7. 150 Best Practices and Tips (Russian Edition)
* "The Aeneid Of Virgil, Translated Into English Verse: By J(ohn) M(yers) King. In 2 Vol; Volume 1"
* Sophronime (- (quatre) Dialogue(s) Par Fénelon)... (French Edition)
* M(y) R(easons) E(xplanations): Crash Course in Contemporary Sociology: An Urban Perspective: Ideas and Disorders of Society (My Reasoned Explanations) (Volume 1)
* Quantitative Analysis and Ibm(r) Spss(r) Statistics: A Guide for Business and Finance (Statistics and Econometrics for Finance)

Some of these were the sort of thing I was trying to catch, but I had to search over 2 million titles to get that small list of examples, so these are not common. I'll add some more test cases as suggested in the PR and try to organise the regexs a bit better so at least we know what cases are handled well, and which could be problematic.
